### ISSUE TITLE
[2.4.4] Missing Frontend CacheInfo breaks bin/magento

### ISSUE THREAD

**Issue by Hexmage** [magento/magento2#35812]
### Preconditions and environment

- Magento 2.4.4
- Upgrade from 2.3.2
- Website is a PWA so doesn't use regular frontend.

### Steps to reproduce

composer update --no-dev
Run bin/magento.

### Expected result

See a list of commands.

### Actual result

```Warning: Undefined array key "frontend" in ../vendor/magento/framework/App/Cache/Frontend/Pool.php on line 90```

### Additional information

I have no clue what causes the exact scenario to start happening. But it appears that it's a minor programming oversight. Where someone assumed that if cacheInfo exists, the frontend cacheInfo also exists which apparently isn't true.

Fix is changing line 89 from
`if (null !== $cacheInfo) {`
to
`if (null !== $cacheInfo && isset($cacheInfo[FrontendPool::KEY_FRONTEND_CACHE]) {`

### Triage and priority

- [X] Severity: **S0** _- Affects critical data or functionality and leaves users without workaround._
- [ ] Severity: **S1** _- Affects critical data or functionality and forces users to employ a workaround._
- [ ] Severity: **S2** _- Affects non-critical data or functionality and forces users to employ a workaround._
- [ ] Severity: **S3** _- Affects non-critical data or functionality and does not force users to employ a workaround._
- [ ] Severity: **S4** _- Affects aesthetics, professional look and feel, “quality” or “usability”._

**Comment by engcom-Delta** [comment]
Hi @Hexmage ,
Thank you for raising an issue , We tried to reproduce the issue on Magento PWA Venia instance and Magento 2.4.4 instance ,Issue is not reproducible. Hence review provided steps.

Please find the attached reference video.
https://www.loom.com/share/7e2490b4c3fc4f5ca6922c16c81d9e62

**Comment by devchris79** [comment]
@Hexmage I just had this issue, I initially tried the fix you suggested but it still reported errors for me.

I got to the root for my instance, we use LiteSpeed cache and when I looked in the admin panel it had dropped back to built in cache (I commented out lines 89-91 of pool.php temporarily). The culprit was somehow our .htaccess was missing this:

```
<IfModule Litespeed>
    LiteMage on
</IfModule>
```
When I added this the cache setting selection had reverted back to Litespeed without me reselecting it.
It would seem just as you found that when the expected cache configuration isn't there it crashes.

**Comment by Hexmage** [comment]
@devchris79 Then I probably know what it was. My staging and live environments run on Varnish. But my develop environment doesn't. But I had synced the config from my staging to my develop environment.

So the real issue is that if a caching method is selected in the configuration that can't work it throws a very awkward error.

**Comment by devchris79** [comment]
@Hexmage It would certainly seem that is the case. I am guessing the data provider function getConfigData isn't retrieving an intended null. Maybe the data array element is there but with invalid elements within the config.  

```
 public function getConfigData()
    {
        return $this->data['config'] ?? [];
    }
```

**Comment by devchris79** [comment]
@Hexmage Do you now have an additional entry in your env.php?
Something like:

,
‘cache’ => [
‘graphql’ => [
‘id_salt’ => ‘SALTHERE’
]
]

**Comment by devchris79** [comment]
@Hexmage I think your original fix is what's required, I think that I had two issues going on at the same time with Litespeed cache loss provoking your issue.

**Comment by richardreen** [comment]
I did change the line in pool.php as described here .
But new error, syntax error....I can not find the error...

syntax error, unexpected '{'#0 /home/user222/domains/xxxxxx.xx/public_html/vendor/composer/ClassLoader.php(428): Composer\Autoload\includeFile('/home/u58658p55...')
#1 [internal function]: Composer\Autoload\ClassLoader->loadClass('Magento\\Framewo...')
#2 /home/user222/domains/xxxxxx.xx/public_html/vendor/magento/framework/ObjectManager/Factory/AbstractFactory.php(121): spl_autoload_call('Magento\\Framewo...')
#3 /home/user222/domains/xxxxxx.xx/public_html/vendor/magento/framework/ObjectManager/Factory/Compiled.php(108): Magento\Framework\ObjectManager\Factory\AbstractFactory->createObject('Magento\\Framewo...', Array)
#4 /home/user222/domains/xxxxxx.xx/public_html/vendor/magento/framework/ObjectManager/Factory/Compiled.php(150): Magento\Framework\ObjectManager\Factory\Compiled->create('Magento\\Framewo...')
#5 /home/user222/domains/xxxxxx.xx/public_html/vendor/magento/framework/ObjectManager/Factory/Compiled.php(79): Magento\Framework\ObjectManager\Factory\Compiled->get('Magento\\Framewo...')
#6 /home/user222/domains/xxxxxx.xx/public_html/vendor/magento/framework/ObjectManager/Factory/Compiled.php(150): Magento\Framework\ObjectManager\Factory\Compiled->create('Magento\\Framewo...')
#7 /home/user222/domains/xxxxxx.xx/public_html/vendor/magento/framework/ObjectManager/Factory/Compiled.php(79): Magento\Framework\ObjectManager\Factory\Compiled->get('Magento\\Framewo...')
#8 /home/user222/domains/xxxxxx.xx/public_html/vendor/magento/framework/ObjectManager/Factory/Compiled.php(150): Magento\Framework\ObjectManager\Factory\Compiled->create('Magento\\Framewo...')
#9 /home/user222/domains/xxxxxx.xx/public_html/vendor/magento/framework/ObjectManager/Factory/Compiled.php(79): Magento\Framework\ObjectManager\Factory\Compiled->get('Magento\\Framewo...')
#10 /home/user222/domains/xxxxxx.xx/public_html/vendor/magento/framework/ObjectManager/ObjectManager.php(70): Magento\Framework\ObjectManager\Factory\Compiled->create('Magento\\Framewo...')
#11 /home/user222/domains/xxxxxx.xx/public_html/vendor/magento/framework/App/ObjectManager/Environment/Compiled.php(113): Magento\Framework\ObjectManager\ObjectManager->get('Magento\\Framewo...')
#12 /home/user222/domains/xxxxxx.xx/public_html/vendor/magento/framework/App/ObjectManagerFactory.php(191): Magento\Framework\App\ObjectManager\Environment\Compiled->configureObjectManager(Object(Magento\Framework\Interception\ObjectManager\Config\Compiled), Array)
#13 /home/user222/domains/xxxxxx.xx/public_html/vendor/magento/framework/App/Bootstrap.php(212): Magento\Framework\App\ObjectManagerFactory->create(Array)
#14 /home/user222/domains/xxxxxx.xx/public_html/vendor/magento/framework/App/Bootstrap.php(127): Magento\Framework\App\Bootstrap->__construct(Object(Magento\Framework\App\ObjectManagerFactory), '/home/user222...', Array)
#15 /home/user222/domains/xxxxxx.xx/public_html/vendor/magento/framework/Console/Cli.php(184): Magento\Framework\App\Bootstrap::create('/home/u58658p55...', Array)
#16 /home/user222/domains/xxxxxx.xx/public_html/vendor/magento/framework/Console/Cli.php(84): Magento\Framework\Console\Cli->initObjectManager()
#17 /home/user222/domains/xxxxxx.xx/public_html/bin/magento(22): Magento\Framework\Console\Cli->__construct('Magento CLI')
#18 {main}

**Comment by devchris79** [comment]
@richardreen, I think @Hexmage has a small error from his suggested fix below, he is missing a closing bracket from the if statement: 

> Fix is changing line 89 from
> if (null !== $cacheInfo) {
> to
> if (null !== $cacheInfo && isset($cacheInfo[FrontendPool::KEY_FRONTEND_CACHE]) {

Try this instead:
 `if (null !== $cacheInfo && isset($cacheInfo[FrontendPool::KEY_FRONTEND_CACHE])) {`

**Comment by richardreen** [comment]
oke, corrected it and now it runs...no error about syntax error anymore. 
Let's see if this is solving the graphql error.

**Comment by richardreen** [comment]
now presenting a new error:
Notice: Undefined index: query in /home/u222/domains/xxxxxx.xx/public_html/vendor/magento/module-graph-ql/Controller/GraphQl.php on line 208

and the salt entry still added to env.php

**Comment by devchris79** [comment]
That issue with a fix is here: #35412

**Comment by richardreen** [comment]
after correcting that, the next error

{"errors":[{"message":"Syntax Error: Unexpected <EOF>","extensions":{"category":"graphql"},"locations":[{"line":1,"column":1}]}]}

and 500 error and salt...in the env.php..
still not solved!!

**Comment by engcom-Delta** [comment]
Hi @Hexmage ,
Are you still facing the same issue ? . Kindly confirm.

Thanks

**Comment by Hexmage** [comment]
@engcom-Delta 
I've implemented my workaround with a patch to avoid the issue. But I'm very confident it's caused by having a caching system configured that doesn't run on your system. eg. Having Varnish configured but don't have any Varnish installed.

**Comment by engcom-Delta** [comment]
Hi @Hexmage ,
Thank you for the update. Shall we close it ? then?

Regards,

**Comment by Hexmage** [comment]
@engcom-Delta 
This scenario could use some better error handling. Because it's very unclear that a incorrectly configured FPC is causing the issue.

**Comment by engcom-Dash** [comment]
Hi @Hexmage,

We are checking the issue with 2.4-develop instance and we are not able to reproduce the issue because we are getting expected result....

Screenshots:
![Screenshot from 2022-09-20 15-02-44](https://user-images.githubusercontent.com/60198592/191231149-92f3cb4e-9bc4-4b7f-895c-b5dd1f89161f.png)
![4](https://user-images.githubusercontent.com/60198592/191231183-51192c38-7ced-46e8-8187-98276017aabf.png)

Steps:
1.install Magento instance
2.Update the instance using composer update command
3.Composer update completed successfully


Let us know if we missing anything!!!

Thanks!!!

**Comment by RajasekarPanneerselvam** [comment]
@magento give me 2.4-develop instance

**Comment by engcom-Dash** [comment]
@magento give  me 2.4-develop instance

**Comment by engcom-Dash** [comment]
@magento give me 2.4-develop instance

**Comment by engcom-Dash** [comment]
Hi @Hexmage ,

As per your instructions we enabled varnish on the server before upgrade...Even though we getting the same results.

Screenshots:
<img width="1512" alt="Screenshot 2022-10-19 at 6 01 28 PM" src="https://user-images.githubusercontent.com/60198592/196691921-24171789-519d-40a9-be02-7183360f4d2d.png">

Let us know if we missing anything!!!

Thanks!

**Comment by Hexmage** [comment]
@engcom-Dash 
I've had the issue happen to multiple different shops between 2.4.4 and 2.4.5. But I have no clue how the effectively reproduce it on a clean installation.

**Comment by engcom-Dash** [comment]
Hi @Hexmage ,

We tried both Local and cloud instance we are same results...Any how we will try once with fresh installation and will update you ASAP.

Thanks!!!

**Comment by Hexmage** [comment]
@engcom-Dash 
We found the cause of the issue, we had the following node in our `app/etc/env.php`
```
    'cache' => [
        'graphql' => [
            'id_salt' => '[A SALT HASH]'
        ]
    ] 
```

**Comment by devchris79** [comment]
@engcom-Dash 
This is the same issue as #35861

**Comment by engcom-Dash** [comment]
Hi @devchris79 ,

We are Closing this issue seems its like duplicate [ https://github.com/magento/magento2/issues/35861]( )

Let us know if we missing anything!!!

Thanks!

**Comment by mreichenbach-igniti** [comment]
@engcom-Dash  We have a Magento 2.4.4-p2 and can reproduce the problem just my calling the "/graphql" URL of the store. This adds the "id_salt" value into the `env.php` and afterwards the frontend returns only http-500. On the other side we could not reproduce this behaviour with Magento 2.4.3. Sounds like a serious problem to me.

I'm confused: this issue was claused with the reason that it's a duplicate of #35861, but #35861 itself is also closed. So no one is working on this at all?

**Comment by richardreen** [comment]
In magento 2.4.5 this error is not thereanymore

Verstuurd vanaf mijn iPad

**Comment by jmonrove** [comment]
Just happened to me on 2.4.5-p1, problem is still there

**Comment by munkhulzii** [comment]
I had this issue also with 2.4.5-p1. 
In system.log I see the error message 
`main.ERROR: Syntax Error: Unexpected <EOF>

GraphQL (1:1)
1:
   ^
 [] []
 `
 and after that the shop is down.

**Comment by munkhulzii** [comment]
Found out, it happens if the env.php does not have the cache -> frontend node. After the cache configuration was added correctly the issue does not appear anymore.

**Comment by Hexmage** [comment]
@engcom-Dash 
We have known the cause of the issue for a while now, but I don't think it's fixed in the latest version yet.

So if your `app/etc/env.php` doesn't contain a `cache -> frontend` node yet (which it doesn't on a new install). And you visit the `/graphql` endpoint. The shop breaks. 

This is caused because the call to the graphql endpoint creates a `cache -> graphql` node in the env. But the `./vendor/magento/framework/App/Cache/Frontend/Pool.php` class expects the frontend node to always exists when the cache node exists. Which clearly isn't the case.

**Comment by erikvandermey** [comment]
I just started encountering this issue on a fresh 2.4.6-p1 build. Hitting the /graphql/ url kills the site.

Short of installing Varnish and enabling cache to get the needed node inserted into env.php, does someone have the snippet of code for env.php that would prevent this error from happening?

**Comment by experius-nl** [comment]
@erikvandermey 
```
'cache' => [
        'frontend' => []
 ]
```
There might already be a different value in the `cache` node (hence why the error is triggering), but adding an empty frontend node in addition to the other nodes already there fixes the issue.

**Comment by hostep** [comment]
Ran into this using Magento 2.4.6-p3
Workaround by adding `frontend => []` works.

I don't understand why this issue got closed though. @engcom-Dash: it got closed because it's a duplicate of #35861 and  #35861 got closed because it's a duplicate of this issue, that makes little sense? One of both should remain open, otherwise it can't be picked up.
This should be a pretty high prio issue (P0 or P1). We just had this happened a couple of hours after a deploy where everything went fine, it's just after some hours when a specific url got hit that we got that `graphql => id_salt` added to the `env.php` file. And that broke the entire shop...

Or has it already been fixed on `2.4-develop` in the mean while? If yes: can somebody share some info or a commit?

**Comment by hostep** [comment]
Okay, dug in a bit.

This probably only happens for shops that were first setup with a version before Magento 2.3.1, and that later got updated to a newer version.

Before 2.3.1, [this change](https://github.com/magento/magento2/pull/18641) wasn't included yet, which writes `cache > frontend > default/page_cache > id_prefix` section to the `env.php` file at **installation** time.
However, if you later upgrade Magento to a newer version, this isn't written to the `env.php` file.
So if we then call the `/graphql` url/endpoint, Magento will write `cache > graphql > id_salt` to the `env.php` file using [this code](https://github.com/magento/magento2/blob/2.4.6-p3/app/code/Magento/GraphQlCache/Model/CacheId/CacheIdCalculator.php#L124-L128).

And the [code mentioned before](https://github.com/magento/magento2/blob/2.4.6-p3/lib/internal/Magento/Framework/App/Cache/Frontend/Pool.php#L89-L91) crashes because it expects a `cache > frontend` section, which isn't always guaranteed to be there.

----

A good fix in my opinion is to run `bin/magento setup:config:set` without any parameters, it will update the `env.php` file with all the missing sections automatically (make sure to make a backup of your `env.php` file so you can compare what got updated). So if you got yourself into the situation this issue describes:

1. First, execute one of these two:
    - remove the entire `cache` section from the `env.php` file
    - or add an empty `frontend => []` to the `cache` section of the `env.php` file
2. Now run `bin/magento setup:config:set`

----

However, we still need to fix this crash that can happen, will send in a PR shortly with a fix.
